### PR TITLE
TST: Test ArrowExtensionArray with decimal types

### LINKED ISSUE #34166: ENH: Add high-performance DecimalArray type?
#### Is your feature request related to a problem?

Doing operations with decimal types is currently very slow, since pandas stores them as python Decimal objects. I think it would be nice to have a high-performance, vectorized DecimalArray ExtensionArray type. Would you be open to a PR adding a type like this, or is this something you think belongs outside the pandas library?

#### Describe the solution you'd like

Create an ExtensionArray datatype that internally represents numbers similar to in pyarrow (as ints, with precision and scale). (Require that all values have the same precision and scale.) Implement basic arithmetic operations (+, -, *, /, etc.).

#### API breaking implications

By adding a `__from_arrow__` and `__to_arrow__` field to DecimalArray, this would cause decimal data from pyarrow to be converted to DecimalArray rather than object array by default, which would be a breaking change.

#### Describe alternatives you've considered

The status quo (loading as object arrays) works, but is slow.

#### Additional context

This would also be really helpful for loading parquet files that decimal types via pyarrow more quickly, since the step to instantiate python Decimal objects for each value can be very slow for large DataFrames.

### PR DESCRIPTION
**Pull request by mroeschke**
closes #34166

### CODE REVIEW

**Comment by jbrockmendel** [comment]
[XPASS(strict)] No pyarrow kernel for decimal128(4, 3)

File: pandas/tests/extension/test_arrow.py
Comment (jbrockmendel) [inline pandas/tests/extension/test_arrow.py:1109]: comment about why comparing directly doesnt work?

Comment (mroeschke) [reply pandas/tests/extension/test_arrow.py:1123]: Good point. Added

File: pandas/tests/extension/test_arrow.py
Comment (jbrockmendel) [inline pandas/tests/extension/test_arrow.py:88]: looks like if you pass Decimal("nan") to pyarrow it raises.  is that going to be relevant for us?

File: pandas/tests/extension/test_arrow.py
Comment (jbrockmendel) [inline pandas/tests/extension/test_arrow.py:267]: why doesn't this one work?

File: pandas/tests/extension/test_arrow.py
Comment (jbrockmendel) [inline pandas/tests/extension/test_arrow.py:366]: i think we agreed to change/fix the ArrowDtype.type behavior.  since its an API change, let's try to get that in for 2.0

File: pandas/tests/extension/test_arrow.py
Comment (jbrockmendel) [inline pandas/tests/extension/test_arrow.py:835]: #50689 fixes this for timestamptz, after both this and that are merged we can see about fixing this for decimal.

File: pandas/tests/extension/test_arrow.py
Comment (mroeschke) [reply pandas/tests/extension/test_arrow.py:88]: Looks like we coerce to NA

```
In [6]: pd.Series([decimal.Decimal("nan")], dtype=pd.ArrowDtype(pa.decimal128(4, 1)))
Out[6]:
0    <NA>
dtype: decimal128(4, 1)[pyarrow]
```

Which I guess is fuzzy with the NA vs nan debate

File: pandas/tests/extension/test_arrow.py
Comment (mroeschke) [reply pandas/tests/extension/test_arrow.py:267]: `pa.type_for_alias` doesn't have an alias for `"decimal128..."` (similar to timestamps w/ tz) so this fails

Comment (jbrockmendel) [reply pandas/tests/extension/test_arrow.py:269]: OK.  Can we get #50689 in and see if we can fix this case too

File: pandas/core/indexes/base.py
Comment (jbrockmendel) [inline pandas/core/indexes/base.py:4956]: can we be more specific than this?  e.g. time32 would also go through here

File: pandas/tests/extension/test_arrow.py
Comment (jbrockmendel) [inline pandas/tests/extension/test_arrow.py:671]: is this something we can patch in construct_from_string (OK for out of scope, just curious)

File: pandas/core/indexes/base.py
Comment (mroeschke) [reply pandas/core/indexes/base.py:4956]: `is_numeric_dtype(self.dtype)` should block time32. I could import pyarrow and check `pa.types.is_decimal` if you prefer`

File: pandas/tests/extension/test_arrow.py
Comment (mroeschke) [reply pandas/tests/extension/test_arrow.py:671]: Yes I believe so

File: pandas/core/indexes/base.py
Comment (jbrockmendel) [reply pandas/core/indexes/base.py:4956]: makes sense, this is fine

**Review by jbrockmendel — APPROVED** [review]
LGTM pending greenish

**Comment by mroeschke** [comment]
Greenish. Merging.

**Pull request merged**
